lib/index.js

```diff
@@ -3,32 +3,48 @@
 const Bluebird = require('bluebird');
 const Joi      = require('joi');
 
-const Validator = require('./validator');
-
-const RADIX = 10;
+const ApproximateCountValidator = require('./validators/approximate-count');
+const Counter                   = require('./counter');
+const Key                       = require('./key');
+const TotalCountValidator       = require('./validators/total-count');
 
 exports.register = (server, options, next) => {
 
   server.ext('onPreResponse', (request, reply) => {
 
     const settings = request.route.settings.plugins.totalCount;
-    const credentials = request.auth.credentials;
-    const validation = Joi.validate(request, Validator, { allowUnknown: true });
+    const approximateCountValid = !Joi.validate(request, ApproximateCountValidator, { allowUnknown: true }).error;
+    const totalCountValid = !Joi.validate(request, TotalCountValidator, { allowUnknown: true }).error;
 
-    if (!settings || validation.error) {
+    if (!settings || !approximateCountValid && !totalCountValid) {
       return reply.continue();
     }
 
-    return Bluebird.resolve(settings.model)
-    .then((Model) => {
-      if (typeof Model.prototype.filter === 'function') {
-        return new Model().filter(request.query.filter, credentials).count('*');
+    const Redis = options.redisClient;
+    const Model = settings.model;
+    const ttl = options.ttl;
+    const key = Key.generate(Model, request, options.uniqueKey);
+
+    return Bluebird.resolve()
+    .then(() => {
+      if (Redis && approximateCountValid) {
+        return Redis.get(key);
       }
-      return new Model().count('*');
+    })
+    .then((cachedCount) => {
+      if (cachedCount) {
+        return parseInt(cachedCount);
+      }
+
+      return Counter.count(Model, request, Redis, key, ttl);
     })
     .then((count) => {
-      request.response.source.total_count = parseInt(count, RADIX);
-      reply.continue();
+      if (approximateCountValid) {
+        request.response.source.approximate_count = count;
+      } else {
+        request.response.source.total_count = count;
+      }
```

Comment:
I thought about that, but product wanted it to be a different field returned. For our purposes, this will be an undocumented feature of our API only used in dashboard too.